core/service.go

```diff
@@ -2,26 +2,542 @@ package core
 
 import (
 	"context"
+	"encoding/json"
 	"fmt"
+	"io"
+	"sort"
 	"strings"
+	"sync"
 	"syscall"
 	"time"
 
+	"github.com/dagger/dagger/core/pipeline"
+	"github.com/dagger/dagger/router"
 	"github.com/moby/buildkit/client/llb"
 	bkgw "github.com/moby/buildkit/frontend/gateway/client"
 	"github.com/moby/buildkit/identity"
 	"github.com/moby/buildkit/solver/pb"
 	"github.com/opencontainers/go-digest"
+	"github.com/pkg/errors"
 	"github.com/vito/progrock"
 	"golang.org/x/sync/errgroup"
 )
 
 type Service struct {
-	Container *Container
-	Detach    func()
+	// SessionID is the ID of the session that created this service.
+	SessionID string `json:"sid"`
```

Comment:
Yeah, that's true. This was a last-ditch effort change to see if it was the root cause for tests failing in parallel, and it fixed it, but definitely at the cost of cache invalidation on every session run.

I'm not sure about DNS trickery, but we're tantalizingly close to being able to just do away with DNS entirely. Now that we're running services via gateway containers it'd be really easy to run an internal shim command to collect the IP. Then it's just a matter of figuring out how to pass that IP to the clients without busting the cache. "Just" is doing a lot of work there -- ideas welcome. 😉

Will think about this more tomorrow!

Reply:
Have all of the scaffolding in place for the above, using secrets as the special sauce for passing the host=>IP mapping to clients, but I forgot that gateway containers can only run one process, so we can't just run a command to get the IP before starting the service. Going to try a little [IP exchange dance](https://github.com/vito/bass/commit/691753f3bf5a00b6d128723b8e68d45597bbe317#diff-5c121f0bd9a6158ecfd3accf04381afe3b1a3228ffe10993011a2b7f23940c46L488-L513) I did in Bass prior to switching to DNS. (The reason for switching to DNS was to avoid cache-busting; I hadn't thought of the secrets trick!)

Reply:
> Edit: I guess I could pass them in as a secret. 🤔

Yeah that's the direction I was getting at in the end of my last comment, though I was imagining we'd still use DNS. Probably different variations possible, but I was thinking something like
* still generating a DNS name for the service based on a hash that includes the session ID
* in dependent execs, the LLB will have a secret env setup for the shim that tells it the aliases to put in /etc/hosts.
   * the secret env name (i.e. the identifier, not actual value) could include the a hash of the parts of the service that we actually want to be part of the exec's cache key (i.e. if the definition of the service container changes, then exec will re-run)
* The shim uses that to set /etc/hosts with aliases and then execs the process

Just one idea though, could be missing something

Reply:
(Sorry, I wrote my last comment, reloaded the page, and then saw your more recent comment)

Reply:
Resolved via https://github.com/moby/buildkit/pull/3960 - we now affix each hostname with a session-scoped string which is configured as the search domain and propagated to child Dagger sessions.